test/mlir/accelerators/nnpa/conversion/onnx-to-zhigh/gemm.mlir

```diff
@@ -74,6 +74,37 @@ func @test_gemm_transAB(%arg0 : tensor<10x5xf32>, %arg1 : tensor<5x10xf32>, %arg
 
 // -----
 
+// COM: Lower ONNXGemm to ZHigh without verifying the actual size of an unknown
+// COM: dimension. Accelerator will emit an error if the dimension size is
```

Comment:
Theoretically, yes. And it would happen to CPU code too, say, you run with a very large model which is beyond the host memory. NNPA supports smaller tensors than CPU does but the problem does not change.

zDNN team thinks zDNN can support large tensors for typical models. Otherwise, zDNN will emit errors. So we rely on that at this moment. This looks similar to what current DL frameworks are supporting GPUs. If the model is very large, it just fails.

> An alternative would be to version the code and check at runtime that the sizes of all unknown dimensions are within the range supported by the accelerator.

Yes, it would a solution. We discussed this lightweight approach at one of our previous local meetings (I remember @caoimhinuibrian mentioned that). 

At this moment, zDNN team thinks zDNN

Reply:
I agree that very large models may exhaust resource on the CPU. The question is whether a model that does not exhaust resource when run on the CPU would fail to run correctly because the compiler decided to optimize it and run some operators on an accelerator.  I see that as a failure of the compiler, which IMO should not have offloaded an operator without first checking the constraints.  

I think the best long term solution would be to version the code to avoid this situation altogether. But, let's say we decide to ignore the NNPA constrain and offload it anyway (speculatively), when things go wrong we should make it possible for a user to recover from the runtime error in some way.  I guess we have a way now:  stop using -maccel=NNPA (but that is obviously a big hammer). And that might be acceptable in the short term. 

> zDNN team thinks zDNN can support large tensors for typical models. Otherwise, zDNN will emit errors.

Compared to what the CPU can handle (in terms of tensor sizes) what is the limit the zDNN has ?  If it is much smaller than the CPU then we need to avoid offloading IMO, if it is close then I would be OK with the "just offload anyway" approach given that for most models, in practice, we would not trip into this problematic scenario.

Reply:
I understand the concerns, @etiotto. Since we are talking about a very rare case (i.e. the dimension is too large for the HW), best is to eventually have an assert, if the zdnn internal code does not already have one (please verify).

In general, we should encourage a "fussy" mode where each malloc and other possible point of failures such as this one are tested. 

In the meanwhile, in case sizes restrictions are not tested in the zndd code, I would simply recommend to add a `// TODO` in the code so that when we handle a fussy mode, we may properly check for all such conditions.

Reply:
> Since we are talking about a very rare case (i.e. the dimension is too large for the HW), best is to eventually have an assert, if the zdnn internal code does not already have one (please verify)

zdnn (NNPA) actually has. If one dimension is too large it will emit an error.